File: reverse.java
// Time Complexity :O(n)
// Space Complexity :O(1)
// Did this code successfully run on Leetcode :yes
// Any problem you faced while coding this :no


// Your code here along with comments explaining your approach
class ListNode {
         int val;
         ListNode next;
         ListNode() {}
         ListNode(int val) { this.val = val; }
         ListNode(int val, ListNode next) { this.val = val; this.next = next; }
     }
public class reverse{
public ListNode reverseList(ListNode head) {
    if(head==null || head.next==null)return head;
    
    ListNode reversed =reverseList(head.next);
    head.next.next=head;
    head.next=null;
    return reversed;
    }
}

File: Reverse.py
#Time complexity:O(N)
#space Complexity:O(1)

# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def reverseList(self, head: Optional[ListNode]) -> Optional[ListNode]:
        prev=None
        curr=head
        while curr:
            next_pointer=curr.next
            curr.next=prev
            prev=curr
            curr=next_pointer
        return prev
        


        